providers/mock/mock.go

```diff
@@ -229,49 +234,87 @@ func (p *MockV0Provider) UpdatePod(ctx context.Context, pod *v1.Pod) error {
 		return err
 	}
 
+	p.mu.Lock()
+
 	p.pods[key] = pod
 	p.notifier(pod)
 
+	p.mu.Unlock()
+
 	return nil
 }
 
-// DeletePod deletes the specified pod out of memory.
-func (p *MockV0Provider) DeletePod(ctx context.Context, pod *v1.Pod) (err error) {
-	ctx, span := trace.StartSpan(ctx, "DeletePod")
+func (p *MockV0Provider) TerminatePod(ctx context.Context, pod *v1.Pod) (err error) {
+	ctx, span := trace.StartSpan(ctx, "TerminatePod")
 	defer span.End()
 
 	// Add the pod's coordinates to the current span.
 	ctx = addAttributes(ctx, span, namespaceKey, pod.Namespace, nameKey, pod.Name)
 
-	log.G(ctx).Infof("receive DeletePod %q", pod.Name)
-
 	key, err := buildKey(pod)
 	if err != nil {
 		return err
 	}
 
-	if _, exists := p.pods[key]; !exists {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	pod, ok := p.pods[key]
+	if !ok {
 		return errdefs.NotFound("pod not found")
 	}
 
+	p.terminatePod(ctx, pod)
+	p.pods[key] = pod
+	p.notifier(pod)
+	return nil
+}
+
+func (p *MockV0Provider) terminatePod(ctx context.Context, pod *v1.Pod) {
 	now := metav1.Now()
-	delete(p.pods, key)
 	pod.Status.Phase = v1.PodSucceeded
-	pod.Status.Reason = "MockProviderPodDeleted"
-
-	for idx := range pod.Status.ContainerStatuses {
+	pod.Status.Reason = "MockProviderPodTermianted"
+	for idx, s := range pod.Status.ContainerStatuses {
 		pod.Status.ContainerStatuses[idx].Ready = false
-		pod.Status.ContainerStatuses[idx].State = v1.ContainerState{
-			Terminated: &v1.ContainerStateTerminated{
-				Message:    "Mock provider terminated container upon deletion",
-				FinishedAt: now,
-				Reason:     "MockProviderPodContainerDeleted",
-				StartedAt:  pod.Status.ContainerStatuses[idx].State.Running.StartedAt,
-			},
+		pod.Status.ContainerStatuses[idx].State.Terminated = &v1.ContainerStateTerminated{
+			Message:    "Mock provider terminated container upon request",
+			FinishedAt: now,
+			Reason:     "MockProviderPodContainerTerminated",
+		}
+		if s.State.Running != nil {
+			pod.Status.ContainerStatuses[idx].State.Terminated.StartedAt = s.State.Running.StartedAt
 		}
 	}
 
-	p.notifier(pod)
+}
+
+// DeletePod deletes the specified pod out of memory.
+func (p *MockV0Provider) DeletePod(ctx context.Context, pod *v1.Pod) (err error) {
+	ctx, span := trace.StartSpan(ctx, "DeletePod")
+	defer span.End()
+
+	// Add the pod's coordinates to the current span.
+	ctx = addAttributes(ctx, span, namespaceKey, pod.Namespace, nameKey, pod.Name)
+
+	log.G(ctx).Infof("receive DeletePod %q", pod.Name)
+
+	key, err := buildKey(pod)
+	if err != nil {
+		return err
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	pod, exists := p.pods[key]
+	if !exists {
+		return errdefs.NotFound("pod not found")
+	}
+
+	if pod.Status.Phase == v1.PodRunning {
```

Comment:
Would it be expected all providers do this?

Reply:
No, I need to account for this.

Reply:
Or more precisely, I need to check if the provider implements TerminatePod (and split TerminatePod from the main interface).

I do expect all providers to set the phase of the pod, though.